stellargraph/core/element_data.py

```diff
@@ -335,29 +381,62 @@ def __init__(self, ids, sources, targets, weights, type_starts):
         self._empty_ilocs = np.array([], dtype=np.uint8)
 
     def _init_directed_adj_lists(self):
-        # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            in_dict.setdefault(tgt, []).append(i)
-            out_dict.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_in_dict = _numpyise(in_dict, dtype=dtype)
-        self._edges_out_dict = _numpyise(out_dict, dtype=dtype)
+        # record the edge ilocs of incoming and outgoing edges
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        def _to_dir_adj_list(arr):
+            neigh_counts = np.bincount(arr, minlength=self.number_of_nodes)
+            neigh_counts = neigh_counts.astype(self._id_index.dtype, copy=False)
```

Comment:
I think this may just be creating an array unnecessarily, because the `cumsum` itself can handle the casting internally, via the `dtype` argument that's already being specified.

```suggestion
```

This applies to `_init_undirected_adj_lists` too.